lib/wasm/WasmMainTemplatePlugin.js

```diff
@@ -23,95 +23,97 @@ function getAllWasmModules(chunk) {
+			const usedName = importedModule && importedModule.isUsed(exportName);
 			if (usedName !== false) {
-				array.push({
-					exportName,
-					usedName,
-					module: dep.module,
-					description: dep.description,
-					direct: dep.onlyDirectImport
-				});
-			}
-		}
-	}
-	const importsCode = [];
-	const waitForPromises = new Map();
-	for (const pair of depsByRequest) {
-		const properties = [];
-		for (const data of pair[1]) {
-			if (data.direct) {
-				const instanceVar = `m${waitForPromises.size}`;
-				waitForPromises.set(
-					instanceVar,
-					`installedWasmModules[${JSON.stringify(data.module.id)}]`
-				);
-				properties.push(
-					`${JSON.stringify(data.exportName)}: ${instanceVar}.exports` +
-						`[${JSON.stringify(data.exportName)}]`
-				);
-			} else {
-				const params = data.description.signature.params.map(
-					(param, k) => "p" + k + param.valtype
-				);
+				const description = dep.description;
+				const direct = dep.onlyDirectImport;
 
-				const result = `__webpack_require__(${JSON.stringify(
-					data.module.id
-				)})[${JSON.stringify(data.usedName)}](${params})`;
+				const index = importIndex++;
 
-				properties.push(
-					Template.asString([
-						`${JSON.stringify(data.exportName)}: function(${params}) {`,
-						Template.indent([`return ${result};`]),
-						"}"
-					])
-				);
+				const propertyName = Template.numberToIdentifer(index);
+
+				if (direct) {
+					const instanceVar = `m${waitForInstances.size}`;
+					waitForInstances.set(instanceVar, importedModule.id);
+					properties.push(
+						`${JSON.stringify(propertyName)}: ${instanceVar}` +
+							`[${JSON.stringify(usedName)}]`
+					);
+				} else {
+					const params = description.signature.params.map(
+						(param, k) => "p" + k + param.valtype
+					);
+
+					const mod = `installedModules[${JSON.stringify(importedModule.id)}]`;
+					const func = `${mod}.exports[${JSON.stringify(usedName)}]`;
+
+					properties.push(
+						Template.asString([
+							`${JSON.stringify(propertyName)}: ` +
+								(importedModule.type.startsWith("webassembly")
+									? `${mod} ? ${func} : `
+									: "") +
+								`function(${params}) {`,
+							Template.indent([`return ${func}(${params});`]),
+							"}"
+						])
+					);
+				}
 			}
 		}
-
-		importsCode.push(
-			Template.asString([
-				`${JSON.stringify(pair[0])}: {`,
-				Template.indent([properties.join(",")]),
-				"}"
-			])
-		);
 	}
 
-	if (waitForPromises.size > 0) {
-		const promises = Array.from(waitForPromises.values()).join(", ");
+	if (waitForInstances.size === 1) {
+		const moduleId = Array.from(waitForInstances.values())[0];
+		const promise = `installedWasmModules[${JSON.stringify(moduleId)}]`;
+		const variable = Array.from(waitForInstances.keys())[0];
+		return Template.asString([
+			`${JSON.stringify(module.id)}: function() {`,
+			Template.indent([
+				`return promiseResolve().then(function() { return ${promise}; }).then(function(${variable}) {`,
```

Comment:
If you don't need to wait for a microtask you could just resolve the `promise` directly.

Reply:
I added a comment into the code